File: android_collector/app/src/main/java/com/collector/pdd/engine/HumanBehavior.kt
package com.collector.pdd.engine

import com.collector.pdd.data.CollectConfig
import kotlinx.coroutines.delay
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

/**
 * 拟人节奏：Beta 分布停顿、偶发走神/长停，对齐桌面 human_behavior.py。
 */
object HumanBehavior {

    data class Profile(
        val delayMin: Double,
        val delayMax: Double,
        val longPauseP: Double,
        val idleP: Double,
        val backscrollP: Double,
        val thinkMin: Double,
        val thinkMax: Double,
        val readMin: Double,
        val readMax: Double,
        val keywordGapMin: Double,
        val keywordGapMax: Double,
        val itemGapMin: Double,
        val itemGapMax: Double,
    )

    fun profile(config: CollectConfig): Profile {
        val level = config.humanLevel.lowercase()
        // 等待区间以 Web 下发为准，不再整体放大超出区间；强度只影响停顿/回滑概率
        val baseMin = config.delayMinMs.toDouble()
        val baseMax = max(baseMin, config.delayMaxMs.toDouble())
        val presets = mapOf(
            "gentle" to doubleArrayOf(0.08, 0.05, 0.10),
            "normal" to doubleArrayOf(0.15, 0.10, 0.18),
            "strict" to doubleArrayOf(0.25, 0.18, 0.28),
        )
        val p = presets[level] ?: presets.getValue("strict")
        return Profile(
            delayMin = baseMin,
            delayMax = baseMax,
            longPauseP = p[0],
            idleP = p[1],
            backscrollP = p[2],
            thinkMin = config.thinkMinMs.toDouble(),
            thinkMax = config.thinkMaxMs.toDouble(),
            readMin = config.readMinMs.toDouble(),
            readMax = config.readMaxMs.toDouble(),
            keywordGapMin = config.keywordGapMinMs.toDouble(),
            keywordGapMax = config.keywordGapMaxMs.toDouble(),
            itemGapMin = config.itemGapMinMs.toDouble(),
            itemGapMax = config.itemGapMaxMs.toDouble(),
        )
    }

    /** 非均匀随机停留（ms）。bias=short|long */
    suspend fun sleepMs(lowMs: Double, highMs: Double, bias: String = "short"): Long {
        val low = max(50.0, lowMs)
        val high = max(low, highMs)
        val span = high - low
        val r = if (bias == "long") beta(2.2, 1.6) else beta(1.4, 2.8)
        var ms = low + span * r
        if (Random.nextDouble() < 0.08) {
            ms *= Random.nextDouble(1.6, 2.8)
        }
        val wait = ms.toLong().coerceAtLeast(50L)
        delay(wait)
        return wait
    }

    /**
     * kind: think/action/read/idle/keyword/item
     */
    suspend fun pause(config: CollectConfig, kind: String = "action"): Long {
        val p = profile(config)
        var low: Double
        var high: Double
        var bias: String
        when (kind) {
            "think" -> {
                low = p.thinkMin; high = p.thinkMax; bias = "short"
            }
            "read" -> {
                low = p.readMin; high = p.readMax; bias = "long"
            }
            "idle" -> {
                // 走神也落在配置区间偏长一侧
                low = (p.delayMin + p.delayMax) / 2.0
                high = p.delayMax
                bias = "long"
            }
            "keyword" -> {
                low = p.keywordGapMin; high = p.keywordGapMax; bias = "long"
            }
            "item" -> {
                low = p.itemGapMin; high = p.itemGapMax; bias = "long"
            }
            else -> {
                low = p.delayMin; high = p.delayMax; bias = "short"
            }
        }
        // 偶发走神：仍夹在配置区间内，最多摸到上限
        if (kind in listOf("action", "item", "read") && Random.nextDouble() < p.idleP) {
            low = (p.delayMin + p.delayMax) / 2.0
            high = p.delayMax
            bias = "long"
        }
        // 强制落在 Web 配置的等待区间内（action/item）
        if (kind == "action" || kind == "item") {
            low = max(low, p.delayMin)
            high = min(high, p.delayMax)
            if (high < low) high = low
        }
        return sleepMs(low, high, bias)
    }

    fun shouldBackscroll(config: CollectConfig): Boolean =
        Random.nextDouble() < profile(config).backscrollP

    fun shouldLongPause(config: CollectConfig): Boolean =
        Random.nextDouble() < profile(config).longPauseP

    /** 触摸坐标抖动 */
    fun jitter(base: Float, span: Float): Float =
        base + Random.nextDouble(-span.toDouble(), span.toDouble()).toFloat()

    fun swipeDurationMs(purpose: String = "list"): Long {
        val (lo, hi) = if (purpose == "detail") 280L to 620L else 320L to 720L
        return Random.nextLong(lo, hi + 1)
    }

    fun detailScrollRounds(): Int = Random.nextInt(3, 7)

    fun listScrollRounds(base: Int = 2): Int =
        max(1, (base * Random.nextDouble(0.7, 1.3)).toInt())

    private fun beta(a: Double, b: Double): Double {
        // 简单近似：用两个 Gamma 构造 Beta；Gamma 用 Marsaglia 简化版
        val x = gammaSample(a)
        val y = gammaSample(b)
        val s = x + y
        return if (s <= 0) 0.5 else min(0.999, max(0.001, x / s))
    }

    private fun nextGaussian(): Double {
        // Box-Muller
        val u1 = Random.nextDouble().coerceAtLeast(1e-12)
        val u2 = Random.nextDouble()
        return Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2)
    }

    private fun gammaSample(shape: Double): Double {
        if (shape < 1.0) {
            return gammaSample(shape + 1.0) * Math.pow(Random.nextDouble(), 1.0 / shape)
        }
        val d = shape - 1.0 / 3.0
        val c = 1.0 / Math.sqrt(9.0 * d)
        while (true) {
            var x: Double
            var v: Double
            do {
                x = nextGaussian()
                v = 1.0 + c * x
            } while (v <= 0)
            v = v * v * v
            val u = Random.nextDouble()
            if (u < 1.0 - 0.0331 * (x * x) * (x * x)) return d * v
            if (Math.log(u) < 0.5 * x * x + d * (1.0 - v + Math.log(v))) return d * v
        }
    }
}

File: human_behavior.py
"""拟人浏览节奏：随机停留、分段滚动、鼠标微动、偶发回看与长停顿。

原则：
- 禁止固定间隔、禁止加载完瞬时读数关闭
- 用非均匀分布（偏向短停 + 偶发长停）模拟真人注意力
- 所有页面操作都走本模块，避免各处硬编码 sleep
"""

from __future__ import annotations

import random
import time
from typing import Any, Optional

from loguru import logger


def _clamp(n: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, n))


def _profile(config: dict[str, Any]) -> dict[str, float]:
    """
    human_level: gentle | normal | strict
    数值越大越慢、越像真人逛店。
    """
    level = str(config.get("human_level") or "strict").lower()
    base_min = int(config.get("delay_min") or 1200)
    base_max = int(config.get("delay_max") or 3500)
    if base_max < base_min:
        base_max = base_min

    presets = {
        "gentle": {"mult": 0.85, "long_pause_p": 0.08, "idle_p": 0.05, "backscroll_p": 0.12},
        "normal": {"mult": 1.0, "long_pause_p": 0.15, "idle_p": 0.10, "backscroll_p": 0.18},
        "strict": {"mult": 1.35, "long_pause_p": 0.22, "idle_p": 0.16, "backscroll_p": 0.25},
    }
    p = presets.get(level, presets["strict"])
    return {
        "delay_min": base_min * p["mult"],
        "delay_max": base_max * p["mult"],
        "long_pause_p": p["long_pause_p"],
        "idle_p": p["idle_p"],
        "backscroll_p": p["backscroll_p"],
        "think_min": float(config.get("think_min_ms") or 600),
        "think_max": float(config.get("think_max_ms") or 2200),
        "read_min": float(config.get("read_min_ms") or 1500),
        "read_max": float(config.get("read_max_ms") or 4500),
        "keyword_gap_min": float(config.get("keyword_gap_min_ms") or 4000),
        "keyword_gap_max": float(config.get("keyword_gap_max_ms") or 12000),
        "item_gap_min": float(config.get("item_gap_min_ms") or 2500),
        "item_gap_max": float(config.get("item_gap_max_ms") or 7000),
    }


def human_sleep_ms(low_ms: float, high_ms: float, *, bias: str = "short") -> float:
    """
    非均匀随机停留。
    bias=short: 多数偏短，偶尔偏长（逛列表）
    bias=long:  多数偏中长（读详情）
    """
    low = max(50.0, float(low_ms))
    high = max(low, float(high_ms))
    span = high - low
    if bias == "long":
        # Beta 分布偏向中后段
        r = random.betavariate(2.2, 1.6)
    else:
        # 多数短停 + 长尾
        r = random.betavariate(1.4, 2.8)
    ms = low + span * r
    # 8% 概率插入一次「走神」长停
    if random.random() < 0.08:
        ms *= random.uniform(1.6, 2.8)
    seconds = ms / 1000.0
    time.sleep(seconds)
    return seconds


def pause(config: dict[str, Any], *, kind: str = "action") -> float:
    """统一暂停入口。kind: think/action/read/idle/keyword/item"""
    p = _profile(config)
    mapping = {
        "think": (p["think_min"], p["think_max"], "short"),
        "action": (p["delay_min"], p["delay_max"], "short"),
        "read": (p["read_min"], p["read_max"], "long"),
        "idle": (p["delay_max"], p["delay_max"] * 2.2, "long"),
        "keyword": (p["keyword_gap_min"], p["keyword_gap_max"], "long"),
        "item": (p["item_gap_min"], p["item_gap_max"], "long"),
    }
    low, high, bias = mapping.get(kind, (p["delay_min"], p["delay_max"], "short"))
    # 偶发 idle
    if kind in ("action", "read") and random.random() < p["idle_p"]:
        low, high, bias = mapping["idle"]
        logger.debug("拟人：插入走神停顿")
    sec = human_sleep_ms(low, high, bias=bias)
    logger.debug("拟人停顿 kind={} {:.2f}s", kind, sec)
    return sec


def micro_mouse(page, config: dict[str, Any] | None = None) -> None:
    """鼠标轻微漂移，避免完全静止。"""
    try:
        vp = page.viewport_size or {"width": 390, "height": 844}
        w, h = int(vp.get("width", 390)), int(vp.get("height", 844))
        x = random.randint(int(w * 0.15), int(w * 0.85))
        y = random.randint(int(h * 0.20), int(h * 0.75))
        steps = random.randint(8, 22)
        page.mouse.move(x, y, steps=steps)
        if random.random() < 0.35:
            page.mouse.move(
                _clamp(x + random.randint(-40, 40), 5, w - 5),
                _clamp(y + random.randint(-30, 30), 5, h - 5),
                steps=random.randint(5, 14),
            )
    except Exception as exc:
        logger.debug("鼠标微动跳过: {}", exc)


def human_scroll(
    page,
    config: dict[str, Any],
    *,
    rounds: Optional[int] = None,
    purpose: str = "list",
) -> None:
    """
    分段滚动：距离不一、中途停顿、偶发回滚，模拟真人刷列表/看详情。
    """
    p = _profile(config)
    if rounds is None:
        if purpose == "detail":
            rounds = random.randint(3, 7)
        else:
            base = int(config.get("list_scroll_times") or 10)
            # 不要每次都滚固定次数
            rounds = max(3, int(base * random.uniform(0.65, 1.15)))

    logger.info("拟人滚动 purpose={} rounds={}", purpose, rounds)
    for i in range(rounds):
        micro_mouse(page, config)
        # 分段小滚，而不是一次滚很大
        bursts = random.randint(2, 4)
        for _ in range(bursts):
            delta = random.randint(180, 520) if purpose == "list" else random.randint(120, 380)
            # 偶发轻微左右抖动（移动端几乎无感，但轨迹更不机械）
            dx = random.randint(-8, 8) if random.random() < 0.2 else 0
            page.mouse.wheel(dx, delta)
            human_sleep_ms(120, 420, bias="short")

        # 偶发向上回看
        if random.random() < p["backscroll_p"]:
            page.mouse.wheel(0, -random.randint(80, 260))
            pause(config, kind="action")
            logger.debug("拟人：回看上一段")

        # 滚完一段后停留「扫一眼」
        if purpose == "detail":
            pause(config, kind="read" if i == 0 else "action")
        else:
            pause(config, kind="action")
            if random.random() < p["long_pause_p"]:
                pause(config, kind="idle")


def before_navigate(config: dict[str, Any]) -> None:
    """跳转页面前先想一下，禁止点完立刻下一跳。"""
    pause(config, kind="think")


def after_page_ready(page, config: dict[str, Any], *, purpose: str = "list") -> None:
    """
    页面 ready 后：先微动 + 阅读停顿，再允许解析。
    禁止 networkidle 后瞬时 evaluate。
    """
    micro_mouse(page, config)
    pause(config, kind="read" if purpose == "detail" else "action")
    if random.random() < 0.4:
        micro_mouse(page, config)


def between_keywords(config: dict[str, Any]) -> None:
    pause(config, kind="keyword")


def between_items(config: dict[str, Any]) -> None:
    pause(config, kind="item")


# 兼容旧调用名
def random_delay(delay_min: int, delay_max: int) -> float:
    return human_sleep_ms(delay_min, delay_max, bias="short")
